E2E Test Updates

iisa opened the pull request:
- Extends base e2e test to be re-applied in other contexts
- Create base models for what it means to be BookReader and Navigation
- Update Travis.yml to point to base test only
- Create test file that runs base tests on archive.org - to be manually run.

### Test steps
- pull down branch
- update all modules, cmd `npm install`
- run command: `npm run test:e2e` -> tests pass successfully
- run command: `npm run test:e2e:base` -> tests pass successfully
- run command: `npm run test:e2e:dev` -> tests pass successfully

- run command `npx testcafe ./tests/e2e/ia-production/ia-prod-base.js` -> base tests pass successfully hitting archive.org

Internal Ticket: https://webarchive.jira.com/browse/WEBDEV-3249

bfalling commented on tests/e2e/ia-production/ia-prod-base.js:
```diff
@@ -0,0 +1,20 @@
+const books = [
+  'birdbookillustra00reedrich', // publicDomain
+  // 'pianoservicingtu00rebl_0', // borrowable,
+  // 'adventuresoftoms00twaiiala' // has chapter markers
```
SUGGESTION: Remove commented-out lines (3 above)?

iisa replied:
Same with this. These items are listed in our BR Dev Guide, and they ultimately should be covered by the end to end tests.  My idea is to pre-load them and as we have time as a community, we will unlock it.

bfalling commented on tests/e2e/ia-production/ia-prod-base.js:
```diff
@@ -0,0 +1,20 @@
+  const BR = new BookReader();
+  const url = `${localURL}${page}`;
+  const fixtureName = `Archive.org BR Base Tests for: ${page}`;
+  // loadTest(fixtureName, url, BR);
```
SUGGESTION: Remove commented-out line?

iisa commented on tests/e2e/base.test.js:
```diff
@@ -1,128 +1,19 @@
+const localURL = 'http://127.0.0.1:8000/BookReaderDemo/';
+const localPages = [
+  'demo-ia-plato.html',
+  // 'demo-ia-olivertwist.html'
```
Going to opt out on this one. I wanted to 'pre-load' some known books we eventually want covered.  In the repo-level, this is one of them..

iisa commented on tests/e2e/helpers/base.js:
```diff
@@ -0,0 +1,183 @@
+
+    const nextButton = Nav.desktop.goNext;
+    await t.click(nextButton);
+    await t.wait( 5000 ); // wait for animation and page flip to happen
```
It 'pauses' for the specified amount of time: `https://devexpress.github.io/testcafe/documentation/test-api/pausing-the-test.html`

I found that 1 second is sufficient for local and production testing.  I updated it and made it a constant 👍

iisa commented on tests/e2e/helpers/base.js:
```diff
@@ -0,0 +1,179 @@
+  test('2up mode - Clicking `Previous page` changes the page', async t => {
+    const { Nav, BRcontainer} = BR;
+    const onLoadBR = BRcontainer.child(0);
+    const initialImages = onLoadBR.child('img');
```
I am opting out of this. It is important to state what we are diffing against at after "click" - onLoadBR is just as it is -> BR on first load.

iisa commented on tests/e2e/models/Navigation.js:
```diff
@@ -0,0 +1,32 @@
+import { Selector } from 'testcafe';
```
No, they will grow because mobile navigation is not yet spec'd and scoped.  I'd like to keep them separate, because we will be extending.

iisa commented on tests/e2e/models/Navigation.js:
```diff
@@ -0,0 +1,32 @@
+ * @class
+ * @classdesc defines Navigation base elements
+ */
+export default class Navigation {
```
No, I would like room for the Navigation to be its own thing - because it is.  And it changes shape depending on screen width.  It will help tremendously as we move to test mobile navigation.

iisa commented on package.json:
```diff
@@ -68,6 +68,7 @@
     "serve-dev": "npx concurrently --kill-others npm:serve-live npm:build-js:watch npm:build-css:watch",
     "test": "jest",
     "test:e2e": "testcafe",
+    "test:e2e:base": "testcafe ./tests/e2e/base.test.js",
```
in `.testcafe.json` - we crawl for all the test files to run.  I wanted to isolate for locally run tests in Travis CI to only point to this.
Now, since I isolated production tests to be manually run, then we can remove this.

iisa replied:
done

iisa commented on tests/e2e/base.test.js:
```diff
@@ -1,128 +1,17 @@
+  const url = `${localURL}${page}`;
+  const fixtureName = `Base Tests for: ${page}`;
+  fixture `${fixtureName}`.page `${url}`;
+  runBaseTests(BR);
```
We are using 2 recipes from Testcafe:
1) "Data driven tests" - basically iterate over a data set and apply tests to them.  `https://devexpress.github.io/testcafe/documentation/recipes/create-data-driven-tests.html`
and 2) "Helpers" - abstracted tests that can be shared `https://devexpress.github.io/testcafe/documentation/recipes/extract-reusable-test-code/create-helpers.html`

It looks like some of Testcafe's magic is that it handles the test queue.
`runBaseTests` is just my local helper function that chunks the tests together that needs to be run per item in data set.  We should be fine?

iisa commented on tests/e2e/base.test.js:
```diff
@@ -1,128 +1,17 @@
+const localURL = 'http://127.0.0.1:8000/BookReaderDemo/';
+const localPages = [
+  'demo-ia-plato.html',
+  // 'demo-ia-olivertwist.html'
```
Yes, as what I thought.  But if failed test runner, and I want to keep it here as a "level to unlock" - I didn't have bandwidth to push this through at this time, but it should definitely pass once we uncomment.

iisa commented on tests/e2e/base.test.js:
```diff
@@ -1,128 +1,17 @@
+  const fixtureName = `Base Tests for: ${page}`;
+  fixture `${fixtureName}`.page `${url}`;
+  runBaseTests(BR);
+})
```
sure, but I do like to store variables 😊 - it makes things easier to read.   I can see the scope difference, but does it really make a difference?

cdrini replied:
Variable names make things easier if they're adding meaning to something whose meaning isn't immediately clear. So something like a complex regex, or some computation (then it's that nice self-documenting code we were talking about). `const br = new BookReader()` doesn't add any real information. It doesn't make a huge difference on any specific place, but creates indirection and requires more things to be kept in the reader's memory. It also makes the order of reading commands a little more difficult. For example seeing `runBaseTests(br)` makes you have to look not only at the definition, but also at the rest of the scope to see if `br` is used anywhere else if you want to understand exactly what's happening. `runBaseTests(new BookReader())` is _entirely_ contained. You know that the BR instance being created is a throw away, and isn't used anywhere except that function.

iisa replied:
done. I will keep it as an iterator, as noted in the Testcafe recipe.  That way, also, each test  block has its own scope.

iisa commented on tests/e2e/helpers/base.js:
```diff
@@ -0,0 +1,179 @@
+ * @param { BookReader } BR - Model
+ */
+const PAGE_FLIP_WAIT_TIME = 1000;
+export function runBaseTests (BR) {
```
ok, i can get behind that.

iisa replied:
done

iisa commented on tests/e2e/helpers/base.js:
```diff
@@ -0,0 +1,179 @@
+    const origImg2Src = await initialImages.nth(-1).getAttribute('src');
+
+    const previousButton = Nav.desktop.goPrevious;
+    await t.click(previousButton);
```
My opinion is opposite, but I will update 🤓

iisa replied:
done

iisa commented on tests/e2e/helpers/base.js:
```diff
@@ -0,0 +1,179 @@
+    await t.wait(PAGE_FLIP_WAIT_TIME); // wait for animation and page flip to happen
+
+    const nextBrState = await Selector('.BRcontainer').child(0);
+    const prevImages = nextBrState.child('img');
```
We have a new selector call because we need to see the changed state that happened in the DOM.  The initial `.BRcontainer` selector points only to the loaded state.

iisa commented on tests/e2e/helpers/base.js:
```diff
@@ -0,0 +1,179 @@
+    const origImg2Src = await initialImages.nth(-1).getAttribute('src');
+
+    const nextButton = Nav.desktop.goNext;
+    await t.click(nextButton);
```
done ✅

iisa commented on tests/e2e/helpers/base.js:
```diff
@@ -0,0 +1,179 @@
+    await t.wait(PAGE_FLIP_WAIT_TIME); // wait for animation and page flip to happen
+
+    const nextBrState = await Selector('.BRcontainer').child(0);
+    const nextImages = nextBrState.child('img');
```
same - to capture new state

iisa commented on tests/e2e/helpers/base.js:
```diff
@@ -0,0 +1,179 @@
+    await t.click(Nav.desktop.mode1Up);
+    const onePageViewContainer = await Selector('.BRpageview');
+    await t.expect(onePageViewContainer.visible).ok();
+    const images = onePageViewContainer.find('.BRpagediv1up');
```
done locally

iisa commented on tests/e2e/ia-production/ia-prod-base.js:
```diff
@@ -0,0 +1,18 @@
+  const fixtureName = `Archive.org BR Base Tests for: ${page}`;
+  fixture `${fixtureName}`.page `${url}`;
+  runBaseTests(BR);
+})
```
going to keep `forEach` to isolate scope per test batch

iisa commented on tests/e2e/models/BookReader.js:
```diff
@@ -0,0 +1,15 @@
+export default class BookReader {
+  constructor () {
+    this.shell = new Selector('.BookReader');
+    this.BRcontainer = new Selector('.BRcontainer');
```
sure, but I will opt out because we `BRcontainer` is the thing we know and love.

iisa commented on tests/e2e/models/BookReader.js:
```diff
@@ -0,0 +1,15 @@
+ * BookReader Model
+ * @class
+ * @classdesc defines BookReader base elements
+ */
```
done!

cdrini replied:
Hmmm, still seeing these; is there a commit missing maybe?

cdrini replied:
(Also ditto for the ones in the other models).

iisa commented on tests/e2e/models/BookReader.js:
```diff
@@ -0,0 +1,15 @@
+  constructor () {
+    this.shell = new Selector('.BookReader');
+    this.BRcontainer = new Selector('.BRcontainer');
+    this.Nav = new Nav();
```
done

iisa commented:
ok @cdrini - hope this captures most of the issues/comments that you addressed.  I think this is a solid pass as to get the tests to be easily extendable.  I would like to hold on others - make tasks even in the new Project for community to pitch in - lmk if that is a solid plan 🤓

iisa commented:
> Yaaaay! Prod tests :) Code looks pretty good; a few things about variable naming / simplifying code depth 👍
> 
> A architecture comment: I'm not crazy over the fact that the main tests are now in a helper file :P I think it's a little confusing. Would it be possible to leave them in `base.test.js`, inside the `for` loop? And let that file use environment args to specify the base testing url as well as the paths?
> 
> Ideally, it would be nice to be able to do:
> 
> ```shell
> npm run test:e2e:local ia-plato-blah.html ia-foo-bar.html
> npm run test:e2e:prod birdbookillustra00reedrich
> ```
> 
> Maybe with something like:
> 
> ```json
> "test:e2e:local": "env BASE_URLS='http://127.0.0.1:8000/BookReaderDemo/' testcafe"
> test:e2e:prod: "env BASE_URLS='https://archive.org/details/ https://archive.org/stream/' testcafe"
> ```
> 
> And in the JS, something like:
> 
> ```js
> const DEFAULT_PROD_PATHS = [
>   'birdbookillustra00reedrich',
>   ...
> ];
> const DEFAULT_LOCAL_PATHS = [
>   'demo-ia-plato.html',
>   ...
> ];
> const baseUrls = (process.env.BASE_URLS || 'http://127.0.0.1:8000/BookReaderDemo/').split(' ');
> let paths = baseUrls[0].includes('archive.org') ? DEFAULT_PROD_PATHS : DEFAULT_LOCAL_PATHS;
> if (process.env.PATHS) paths = process.env.PATHS.split(' ');
> ```
> 
> Not sure though; thoughts?

@cdrini these are amazing thoughts - and I think we should capture this in a ticket to update in the future!
